packages/expect/src/utils.ts

```diff
@@ -268,16 +268,36 @@ export const subsetEquality = (
   object: any,
   subset: any,
 ): undefined | boolean => {
-  if (!isObjectWithKeys(subset)) {
-    return undefined;
-  }
+  // subsetEquality needs to keep track of the references
+  // it has already visited to avoid infinite loops in case
+  // there are circular references in the subset passed to it.
+  const subsetEqualityWithContext = (
+    seenReferences: WeakMap<object, boolean> = new WeakMap(),
+  ) => (object: any, subset: any): undefined | boolean => {
+    if (!isObjectWithKeys(subset)) {
+      return undefined;
+    }
 
-  return Object.keys(subset).every(
-    key =>
-      object != null &&
-      hasOwnProperty(object, key) &&
-      equals(object[key], subset[key], [iterableEquality, subsetEquality]),
-  );
+    return Object.keys(subset).every(key => {
+      if (isObjectWithKeys(subset[key])) {
+        if (seenReferences.get(subset[key])) {
```

Comment:
Does this condition need same guard `object != null && hasOwnProperty(object, key) && …` as below?

Because it has been a while since I looked at this function, I am asking, not telling ;)

Reply:
I'm not sure if I fully understand your question, but you have indeed found an edge-case.

If you're wondering whether we should have the same checks in L283 alongside the `.get` the answer is that we only need to have `hasOwnProperty`.

The only reason I check if `subset[key]` is an object is that I want to keep track of all visited objects the first time I see them (L286) and run a comparison straight-away without recursing on `subsetEqualityWithContext` if they have to avoid an infinite loop (L284). Since we will only fall into L284 if `subset[key]` is an `object` there is no need for extra checks such as `object != null`, but it's worth running ` hasOwnProperty(object, key)` to avoid matching against properties up in the prototype chain.

I'm gonna push a fix for this and add a test for this case. Excellent catch, @pedrottimark!

Reply:
Sorry that the line where I wrote the comment is not clear, now that I look at it again :)

What I didn’t know is `object != null` needed to protect the new call `equals(object[key], …)` like the call below?

Looking at `object != null` condition reminds me to say that if you are interested in a follow up pull request for another open issue on this matcher which has an interesting semantic question, see #8308

Reply:
I'm not sure if I fully understand your question, but that's a very good one.

If you're wondering whether we should have the same checks in L283 alongside the `.get` the answer is no. It's not very straightforward to notice this, I have done a couple of checks myself to make sure I was not missing anything too.

The only reason I check if `subset[key]` is an object is that I want to keep track of all visited objects the first time I see them (L286) and run a comparison straight-away without recursing on `subsetEqualityWithContext` if they have to avoid an infinite loop (L284). Since we will only fall into L284 if `subset[key]` is an `object` there is no need for extra checks such as `object != null`. We also don't need to run `hasOwnProperty(object, key)` in L284 because if the property we're checking comes from somewhere else in the prototype chain it will be caught in L293 and therefore we won't ever reach L284 since that only runs if the object we're testing has been seen before.

You can check this by trying, for example:

```
it('does not match properties up in the prototype chain', () => {
  const a = {};
  a.ref = a;

  const b = Object.create(a);
  b.other = 'child';

  const matcher = {other: 'child'};
  matcher.ref = matcher;

  jestExpect(b).not.toMatchObject(matcher);
  expect(() =>
    jestExpect(b).toMatchObject(matcher),
  ).toThrowErrorMatchingSnapshot();
});
```

I have added that very test-case to this PR to avoid possible regressions. Worth noticing that its snapshot seems correct, but if people try matching against a property which comes from a prototype it might be a bit harder for them to figure that out since it will appear when they access it directly but will not appear in the diff. However, IMO, that's expected since logging all properties upwards in the prototype chain would cause way too much noise.

Very good review btw @pedrottimark, thank you very much 💖

Reply:
> since that only runs if the object we're testing has been seen before

Aha, super. Your patience to explain why you do what you do is outstanding.

Reply:
All the brilliant reviews and all the help I have received so far are truly outstanding too, I'm glad I can help. Thanks for the compliment @pedrottimark 😊

(Btw, one of the previous builds seems to have run out of memory. If I'm not mistaken that might be related to the recent issues with memory leaks, so I've just re-triggered the builds)